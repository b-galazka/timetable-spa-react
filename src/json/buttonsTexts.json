{
    "class": "Klasy",
    "teacher": "Nauczyciele",
    "classroom": "Sale"
}