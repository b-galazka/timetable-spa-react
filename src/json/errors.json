{
    "fetchingError": "Wystąpił błąd podczas pobierania danych. Spróbuj ponownie później.",
    "notFound": "Podstrona o tym adresie nie istnieje."
}