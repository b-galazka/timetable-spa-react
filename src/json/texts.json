{
    "lastModified": "Ostatnia aktualizacja",
    "classroom": "sala",
    "schoolClass": "klasa",
    "title": {
        "base": "ZSCKP",
        "notFound": "błąd 404"
    },
    "number": "Nr",
    "hour": "Godz",
    "days": [
        "Poniedziałek",
        "Wtorek",
        "Środa",
        "Czwartek",
        "Piątek"
    ],
    "downloadApp": "Pobierz aplikację mobilną",
    "visitPage": "Przejdź do witryny",
    "goHomepage": "Przejdź na stronę główną"
}