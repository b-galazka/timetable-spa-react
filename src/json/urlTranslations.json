{
    "nauczyciel": "teacher",
    "klasa": "class",
    "sala": "classroom",
    "types": ["nauczyciel", "klasa", "sala"],
    "teacher": "nauczyciel",
    "class": "klasa",
    "classroom": "sala"
}