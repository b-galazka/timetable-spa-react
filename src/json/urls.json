{
    "defaultRedirection": "/klasa/4ET",
    "apiBaseUrl": "http://localhost:4000"
}