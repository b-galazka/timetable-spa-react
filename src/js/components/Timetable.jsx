import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import propTypes from 'prop-types';

import TimetableLoader from './TimetableLoader';
import TimetableHeader from './TimetableHeader';
import TimetableContent from './TimetableContent';

import texts from '../../json/texts';
import urlTranslations from '../../json/urlTranslations';

import '../../scss/timetable.scss';

function mapStateToProps(state) {

    const { fetched, notExists, fetching } = state.timetableObject;

    return {

        timetableObject: {
            fetched,
            fetching,
            notExists
        },

        teachers: state.teachers
    };
}

class Timetable extends Component {

    render() {

        const { fetched, notExists } = this.props.timetableObject;

        return (
            <section className="timetable">

                {
                    ! notExists && <TimetableHeader />
                }

                {
                    fetched ?

                        <TimetableContent /> :

                        <TimetableLoader />
                }

            </section>
        );
    }

    componentDidMount() {

        this.updateTitle();
    }

    componentDidUpdate() {

        this.updateTitle();
    }

    updateTitle() {

        const urlParam = this.props.match.params.type;
        const { fetched, notExists, fetching } = this.props.timetableObject;

        if (notExists) {

            Timetable.setTitle(texts.title.notFound);
        } else if (fetching || fetched) {

            const timetableObjectType = urlTranslations[urlParam];
            const slug = this.getSlug();

            if (timetableObjectType === 'teacher') {

                const { teachers } = this.props;

                const currentTeacher = teachers.find(
                    teacher => teacher.slug === slug
                );

                Timetable.setTitle(currentTeacher.name || currentTeacher.slug);
            } else {

                Timetable.setTitle(slug);
            }
        }
    }

    static setTitle(title) {

        document.title = `${texts.title.base} | ${title}`;
    }

    getSlug() {

        return decodeURIComponent(this.props.match.params.slug);
    }
}

Timetable.propTypes = {

    // redux
    timetableObject: propTypes.shape({
        fetched: propTypes.bool.isRequired,
        fetching: propTypes.bool.isRequired,
        notExists: propTypes.bool.isRequired
    }).isRequired,

    teachers: propTypes.arrayOf(
        propTypes.shape({
            slug: propTypes.string.isRequired,
            name: propTypes.string
        }).isRequired
    ).isRequired,

    // router
    match: propTypes.shape({
        params: propTypes.shape({
            type: propTypes.string.isRequired,
            slug: propTypes.string.isRequired
        }).isRequired
    }).isRequired
};

export default withRouter(connect(mapStateToProps)(Timetable));