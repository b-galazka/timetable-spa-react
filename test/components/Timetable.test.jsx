import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import renderer from 'react-test-renderer';

import Timetable from '../../src/js/components/Timetable';
import { title } from '../../src/json/texts';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('../../src/js/components/TimetableHeader', () => 'TimetableHeader');
jest.mock('../../src/js/components/TimetableContent', () => 'TimetableContent');
jest.mock('../../src/js/components/TimetableLoader', () => 'TimetableLoader');
jest.mock('../../src/json/urlTranslations', () => ({ teacher: 'teacher' }));

jest.mock(
    '../../src/json/texts',
    () => ({
        title: {
            base: 'base title',
            notFound: 'not found title'
        }
    })
);

describe('Timetable component', () => {

    let timetable;

    beforeEach(() => {

        timetable = shallow(
            <Timetable.WrappedComponent

                teachers={[
                    { slug: 'AB', name: 'AB fullname' },
                    { slug: 'BC', name: null },
                    { slug: '#XY', name: 'XY fullname' }
                ]}

                timetableObject={{
                    fetched: false,
                    fetching: false,
                    notExists: false
                }}

                match={{
                    params: { slug: 'XYZ', type: 'teacher' }
                }}
            />
        ).instance();

        timetable.props = { ...timetable.props };
    });

    it(
        'should render timetable header and timetable content if data has been fetched',
        () => {

            const tree = renderer.create(
                <Timetable.WrappedComponent

                    teachers={[{ slug: 'XYZ', name: null }]}

                    timetableObject={{
                        fetched: true,
                        fetching: false,
                        notExists: false
                    }}

                    match={{
                        params: { slug: 'XYZ', type: 'teacher' }
                    }}
                />
            ).toJSON();

            expect(tree).toMatchSnapshot();
        }
    );

    it(
        'should render timetable header and timetable loader if data is being fetched',
        () => {

            const tree = renderer.create(
                <Timetable.WrappedComponent

                    teachers={[{ slug: 'XYZ', name: null }]}

                    timetableObject={{
                        fetched: false,
                        fetching: true,
                        notExists: false
                    }}

                    match={{
                        params: { slug: 'XYZ', type: 'teacher' }
                    }}
                />
            ).toJSON();

            expect(tree).toMatchSnapshot();
        }
    );

    it('should render only timetable loader if data does not exist', () => {

        const tree = renderer.create(
            <Timetable.WrappedComponent

                teachers={[{ slug: 'XYZ', name: null }]}

                timetableObject={{
                    fetched: false,
                    fetching: false,
                    notExists: true
                }}

                match={{
                    params: { slug: 'XYZ', type: 'teacher' }
                }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    describe('Timetable.prototype.componentDidMount', () => {

        it('should set 404 title if timetable object does not exist', () => {

            timetable.props.timetableObject.notExists = true;

            timetable.componentDidMount();

            expect(document.title).toBe(`${title.base} | ${title.notFound}`);
        });

        it('should set title to teacher\'s name if this name exists', () => {

            timetable.props.timetableObject.fetched = true;

            timetable.props.match.params = {
                type: 'teacher',
                slug: 'AB'
            };

            timetable.componentDidMount();

            expect(document.title).toBe(`${title.base} | AB fullname`);
        });

        it('should set title to teacher\'s slug if name does not exist', () => {

            timetable.props.timetableObject.fetched = true;

            timetable.props.match.params = {
                type: 'teacher',
                slug: 'BC'
            };

            timetable.componentDidMount();

            expect(document.title).toBe(`${title.base} | BC`);
        });

        it('should set title to slug from URL if current object is not a teacher', () => {

            timetable.props.timetableObject.fetched = true;

            timetable.props.match.params = {
                type: 'class',
                slug: 'XYZ'
            };

            timetable.componentDidMount();

            expect(document.title).toBe(`${title.base} | XYZ`);
        });

        it('should encode slug from URL', () => {

            timetable.props.timetableObject.fetched = true;

            timetable.props.match.params = {
                type: 'class',
                slug: '%23XY'
            };

            timetable.componentDidMount();

            expect(document.title).toBe(`${title.base} | #XY`);
        });

        it('should encode teacher\'s slug', () => {

            timetable.props.timetableObject.fetched = true;

            timetable.props.match.params = {
                type: 'teacher',
                slug: '%23XY'
            };

            timetable.componentDidMount();

            expect(document.title).toBe(`${title.base} | XY fullname`);
        });
    });

    afterEach(() => {

        document.title = 'default title';
    });
});