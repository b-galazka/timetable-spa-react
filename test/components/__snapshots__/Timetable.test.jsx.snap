// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timetable component should render only timetable loader if data does not exist 1`] = `
<section
  className="timetable"
>
  <TimetableLoader />
</section>
`;

exports[`Timetable component should render timetable header and timetable content if data has been fetched 1`] = `
<section
  className="timetable"
>
  <TimetableHeader />
  <TimetableContent />
</section>
`;

exports[`Timetable component should render timetable header and timetable loader if data is being fetched 1`] = `
<section
  className="timetable"
>
  <TimetableHeader />
  <TimetableLoader />
</section>
`;
